src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -206,53 +254,214 @@ public static FormulaValue Text(EvalVisitor runner, EvalVisitorContext context,
+        {
+            format = Regex.Replace(format, "[hH]", "H");
+            format = Regex.Replace(format, "t+", "");
+
+            return format.Trim();
+        }
+
+        private static string ResolveDateTimeFormatAmbiguities(string format, DateTime dateTime, CultureInfo culture, CancellationToken cancellationToken)
+        {
+            format = ReplaceDoubleQuotedStrings(format, out var replaceList, cancellationToken);        
+            format = TokenizeDatetimeFormat(format, cancellationToken);
+            format = DetokenizeDatetimeFormat(format, dateTime, culture);
+            format = RestoreDoubleQuotedStrings(format, replaceList, cancellationToken);
 
             return format;
         }
 
-        private static string ReplaceWith24HourClock(string format)
+        private static string RestoreDoubleQuotedStrings(string format, List<string> replaceList, CancellationToken cancellationToken)
+        {
+            var stringReplaceRegex = new Regex("\u0011");
+            var array = replaceList.ToArray();
+            var index = 0;
+
+            var match = stringReplaceRegex.Match(format);
+
+            while (match.Success)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                format = format.Substring(0, match.Index) + array[index++].Replace("\"", string.Empty) + format.Substring(match.Index + match.Length);
+                match = stringReplaceRegex.Match(format);
+            }
+
+            return format;
+        }
+
+        private static string ReplaceDoubleQuotedStrings(string format, out List<string> replaceList, CancellationToken cancellationToken)
+        {
+            var ret = string.Empty;
+
+            replaceList = new List<string>();
+
+            foreach (Match match in _internalStringRegex.Matches(format))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                replaceList.Add(match.Value);
+            }
+
+            return _internalStringRegex.Replace(format, "\u0011");
+        }
+
+        private static string DetokenizeDatetimeFormat(string format, DateTime dateTime, CultureInfo culture)
+        {
+            var hasAmPm = format.Contains('\u0001') || format.Contains('\u0002');
+
+            // Day component            
+            format = _daysDetokenizeRegex.Replace(format, dateTime.ToString("dddd", culture))
+                          .Replace("\u0004\u0004\u0004", dateTime.ToString("dddd", culture).Substring(0, 3))
+                          .Replace("\u0004\u0004", dateTime.ToString("dd", culture))
+                          .Replace("\u0004", dateTime.ToString("%d", culture));
+
+            // Month component
+            format = _monthsDetokenizeRegex.Replace(format, dateTime.ToString("MMMM", culture))
+                          .Replace("\u0003\u0003\u0003", dateTime.ToString("MMMM", culture).Substring(0, 3))
```

Comment:
>dateTime.ToString("MMMM", culture).Substring(0, 3) [](http://example.com/codeflow?start=57&length=50)

`dateTime.ToString("MMM", culture)` - DateTime.ToString already handles abbreviated months

Reply:
Yes, I just wanted to be as much close to JS implementation as possible.